# synphage

Pipeline to create phage genome synteny graphics from genbank files

This library provides an intuitive tool to create synteny graphics highlighting the conserved genes between multiple genome sequences.  
This tools is primarily designed to work with phage genomes or other short sequences of interest, although it works with bacterial genomes as well.

Despite numerous synteny tools available on the market, this tool has been conceived because none of the available tools allows to visualise gene conservation in multiple sequences at one glance (as typically cross-links are drawn only between two consecutive sequences for a better readability).

As a result `synphage` was born. 

In addition to show conserved genes across multiple sequences, the originality of this library stands in the fact that when working on the same set of genomes the initial blast and computation need to be run only once. Multiple graphics can then be generated from these data, comparing all the genomes or only a set of genomes from the analysed dataset. Moreover, the generated data is also available to the user as a table, where individual genes or groups of genes can easily be checked by name for conservation or uniqueness in the other genomes.


## Stats 
[![PyPI version](https://badge.fury.io/py/synphage.svg)](https://badge.fury.io/py/synphage)
[![ci](https://github.com/vestalisvirginis/synphage/actions/workflows/ci.yml/badge.svg?branch=main)](https://github.com/vestalisvirginis/synphage/actions/workflows/ci.yml)
[![](https://img.shields.io/pypi/dm/synphage.svg?style=popout-square)](https://pypi.org/project/synphage/)
[![License](https://img.shields.io/github/license/vestalisvirginis/synphage.svg?style=popout-square)](https://opensource.org/licenses/Apache-2.0)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)


## Install

`Synphage` is available via pip install or as docker image.


### Via pip 
```bash
pip install synphage
```


### Via docker
```bash
docker pull vestalisvirginis/synphage:latest
```


### Additional dependencies

Synphage relies on two non-python dependencies that need to be manually installed when Synphage is installed with pip:
- [Blast+](https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/) >= 2.12.0   
- [OpenJDK](https://openjdk.org/projects/jdk/17/) == 17  


## Usage

### Setup 

Synphage requires:
- to specify a folder path where the  `genbank` folder will be present and where generated data will be stored;
- a `genbank` folder populated with genbank files (`.gb` and  `.gbk` extension are accepted);
- a `sequences.csv` file containing the file name and orientation of the sequences to plot.


**Warning**: Genbank file names should not contain spaces.


#### Path setup

```bash
export PHAGY_DIRECTORY=<path_to_data_folder>
```

**Note:** For docker users, this path is defaulted to `/data`.

#### CSV file

```txt
genome_1.gb,0
genome_2.gb,1
genome_3.gb,0
```


### Running Synphage

`Synphage` uses [Dagster](https://dagster.io). In order to run synphage jobs, you need to start dagster first.


#### Starting Dagster

Set up the environment variable DAGSTER_HOME in order to keep a trace of your previous run. For more information, see [Dagster documentation](https://docs.dagster.io/deployment/dagster-instance). 

```bash
export DAGSTER_HOME=<dagster_home_directory>

dagster dev -h 0.0.0.0 -p 3000 -m synphage
```


#### Running the jobs

The current software is structured in three different jobs.
 - `blasting_job` : create the blastn of each sequences against each sequences (results -> gene_identity folder)
 - `transform` : create three tables from the blastn results and genbank files (results -> tables)
 - `synteny_job` : create the synteny graph (results -> synteny)

**Note:** Different synteny plots can be generated from the same set of genomes. In this case the two first jobs only need to be run once and the third job (`synteny_job`) can be triggered separately for each graphs.


## Output

Synphage's output consists of three main parquet files and the synteny graph. However all the data generated by the synphage pipeline are made available in your workng directory.

### Generated data architecture

```
.
â”œâ”€â”€ <path_to_data_folder>/
â”‚   â”œâ”€â”€ genbank/
â”‚   â”œâ”€â”€ fs/
â”‚   â”œâ”€â”€ gene_identity/
â”‚   â”‚   â”œâ”€â”€ fasta/
â”‚   â”‚   â”œâ”€â”€ blastn_database/
â”‚   â”‚   â””â”€â”€ blastn/
â”‚   â”œâ”€â”€ tables/
â”‚   â”‚   â”œâ”€â”€ blastn.parquet
â”‚   â”‚   â”œâ”€â”€ locus_and_gene.parquet
â”‚   â”‚   â””â”€â”€ uniqueness.parquet
â”‚   â””â”€â”€ synteny/
â”‚      â”œâ”€â”€ colour_table.parquet
â”‚      â””â”€â”€ synteny_graph.svg
â””â”€â”€ ...
```


### Tables

The `tables` folder contains the three main parquet files generated by the `transform` job of synphage.   
1. `blastn.parquet` contains the collection of the best match for each locus tag/gene against each genomes. The percentage of identity between two genes/loci are then used for calculating the plot cross-links between the sequences.
1. `locus_and_gene.parquet` contains the full list of `locus tag` and corresponding `gene` names when available for all the genomes in the genbank folder. If the genbank file only contains `CDS`, the locus tag and gene value are replaced by the protein identifyer `protein_id`.
1. `uniqueness.parquet` combined both previous data tables in one, allowing the user to quickly know how many matches their gene(s) of interest has/have retrieved. These data are then used to compute the colour code used for the synteny plot. The result of the computation is recorded in the `colour_table.parquet`. This file is over-written between each `synteny_job` run. 


### Synteny plot

The `synteny plot` is generated as `.svg file` and contains the sequences indicated in the sequences.csv file. The genes are colour-coded according to their abundance (percentage) among the plotted sequences. The cross-links between each consecutive sequence indicates the percentage or similarities between those two sequences.


## Roadmap

- [ ] create config options for the plot at run time
- [ ] create possibility to add ref sequence with special colour coding
- [ ] integrate the NCBI search asset
- [ ] create interactive plot 
- [ ] Help us in a discussion?


## Contributing 

We accept different types of contributions, including some that don't require you to write a single line of code. For detailed instructions on how to get started with our project, see [CONTRIBUTING](CONTRIBUTING.md) file.


## Authors
- [vestalisvirginis](https://github.com/vestalisvirginis) / Virginie Grosboillot / ğŸ‡«ğŸ‡· 


## License
Apache License 2.0
Free for commercial use, modification, distribution, patent use, private use.
Just preserve the copyright and license.


> Made with â¤ï¸ in Ljubljana ğŸ‡¸ğŸ‡®