{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (10, 12)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>id</th><th>timestamp</th><th>check</th><th>level</th><th>column</th><th>rule</th><th>value</th><th>rows</th><th>violations</th><th>pass_rate</th><th>pass_threshold</th><th>status</th></tr><tr><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>i64</td><td>i64</td><td>f64</td><td>f64</td><td>str</td></tr></thead><tbody><tr><td>1</td><td>&quot;2024-06-16 16:12:17&quot;</td><td>&quot;cuallee.check&quot;</td><td>&quot;WARNING&quot;</td><td>&quot;gene&quot;</td><td>&quot;is_complete&quot;</td><td>&quot;N/A&quot;</td><td>4082</td><td>0</td><td>1.0</td><td>1.0</td><td>&quot;PASS&quot;</td></tr><tr><td>2</td><td>&quot;2024-06-16 16:12:17&quot;</td><td>&quot;cuallee.check&quot;</td><td>&quot;WARNING&quot;</td><td>&quot;locus_tag&quot;</td><td>&quot;is_complete&quot;</td><td>&quot;N/A&quot;</td><td>4082</td><td>0</td><td>1.0</td><td>1.0</td><td>&quot;PASS&quot;</td></tr><tr><td>3</td><td>&quot;2024-06-16 16:12:17&quot;</td><td>&quot;cuallee.check&quot;</td><td>&quot;WARNING&quot;</td><td>&quot;cds_gene&quot;</td><td>&quot;is_complete&quot;</td><td>&quot;N/A&quot;</td><td>4082</td><td>0</td><td>1.0</td><td>1.0</td><td>&quot;PASS&quot;</td></tr><tr><td>4</td><td>&quot;2024-06-16 16:12:17&quot;</td><td>&quot;cuallee.check&quot;</td><td>&quot;WARNING&quot;</td><td>&quot;cds_locus_tag&quot;</td><td>&quot;is_complete&quot;</td><td>&quot;N/A&quot;</td><td>4082</td><td>0</td><td>1.0</td><td>1.0</td><td>&quot;PASS&quot;</td></tr><tr><td>5</td><td>&quot;2024-06-16 16:12:17&quot;</td><td>&quot;cuallee.check&quot;</td><td>&quot;WARNING&quot;</td><td>&quot;protein_id&quot;</td><td>&quot;is_complete&quot;</td><td>&quot;N/A&quot;</td><td>4082</td><td>0</td><td>1.0</td><td>1.0</td><td>&quot;PASS&quot;</td></tr><tr><td>6</td><td>&quot;2024-06-16 16:12:17&quot;</td><td>&quot;cuallee.check&quot;</td><td>&quot;WARNING&quot;</td><td>&quot;gene&quot;</td><td>&quot;is_unique&quot;</td><td>&quot;N/A&quot;</td><td>4082</td><td>496</td><td>0.878491</td><td>1.0</td><td>&quot;FAIL&quot;</td></tr><tr><td>7</td><td>&quot;2024-06-16 16:12:17&quot;</td><td>&quot;cuallee.check&quot;</td><td>&quot;WARNING&quot;</td><td>&quot;locus_tag&quot;</td><td>&quot;is_unique&quot;</td><td>&quot;N/A&quot;</td><td>4082</td><td>0</td><td>1.0</td><td>1.0</td><td>&quot;PASS&quot;</td></tr><tr><td>8</td><td>&quot;2024-06-16 16:12:17&quot;</td><td>&quot;cuallee.check&quot;</td><td>&quot;WARNING&quot;</td><td>&quot;cds_gene&quot;</td><td>&quot;is_unique&quot;</td><td>&quot;N/A&quot;</td><td>4082</td><td>496</td><td>0.878491</td><td>1.0</td><td>&quot;FAIL&quot;</td></tr><tr><td>9</td><td>&quot;2024-06-16 16:12:17&quot;</td><td>&quot;cuallee.check&quot;</td><td>&quot;WARNING&quot;</td><td>&quot;cds_locus_tag&quot;</td><td>&quot;is_unique&quot;</td><td>&quot;N/A&quot;</td><td>4082</td><td>0</td><td>1.0</td><td>1.0</td><td>&quot;PASS&quot;</td></tr><tr><td>10</td><td>&quot;2024-06-16 16:12:17&quot;</td><td>&quot;cuallee.check&quot;</td><td>&quot;WARNING&quot;</td><td>&quot;protein_id&quot;</td><td>&quot;is_unique&quot;</td><td>&quot;N/A&quot;</td><td>4082</td><td>45</td><td>0.988976</td><td>1.0</td><td>&quot;FAIL&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (10, 12)\n",
       "┌─────┬────────┬────────┬────────┬────────┬────────┬───────┬──────┬────────┬───────┬───────┬───────┐\n",
       "│ id  ┆ timest ┆ check  ┆ level  ┆ column ┆ rule   ┆ value ┆ rows ┆ violat ┆ pass_ ┆ pass_ ┆ statu │\n",
       "│ --- ┆ amp    ┆ ---    ┆ ---    ┆ ---    ┆ ---    ┆ ---   ┆ ---  ┆ ions   ┆ rate  ┆ thres ┆ s     │\n",
       "│ i64 ┆ ---    ┆ str    ┆ str    ┆ str    ┆ str    ┆ str   ┆ i64  ┆ ---    ┆ ---   ┆ hold  ┆ ---   │\n",
       "│     ┆ str    ┆        ┆        ┆        ┆        ┆       ┆      ┆ i64    ┆ f64   ┆ ---   ┆ str   │\n",
       "│     ┆        ┆        ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆ f64   ┆       │\n",
       "╞═════╪════════╪════════╪════════╪════════╪════════╪═══════╪══════╪════════╪═══════╪═══════╪═══════╡\n",
       "│ 1   ┆ 2024-0 ┆ cualle ┆ WARNIN ┆ gene   ┆ is_com ┆ N/A   ┆ 4082 ┆ 0      ┆ 1.0   ┆ 1.0   ┆ PASS  │\n",
       "│     ┆ 6-16   ┆ e.chec ┆ G      ┆        ┆ plete  ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│     ┆ 16:12: ┆ k      ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│     ┆ 17     ┆        ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│ 2   ┆ 2024-0 ┆ cualle ┆ WARNIN ┆ locus_ ┆ is_com ┆ N/A   ┆ 4082 ┆ 0      ┆ 1.0   ┆ 1.0   ┆ PASS  │\n",
       "│     ┆ 6-16   ┆ e.chec ┆ G      ┆ tag    ┆ plete  ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│     ┆ 16:12: ┆ k      ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│     ┆ 17     ┆        ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│ 3   ┆ 2024-0 ┆ cualle ┆ WARNIN ┆ cds_ge ┆ is_com ┆ N/A   ┆ 4082 ┆ 0      ┆ 1.0   ┆ 1.0   ┆ PASS  │\n",
       "│     ┆ 6-16   ┆ e.chec ┆ G      ┆ ne     ┆ plete  ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│     ┆ 16:12: ┆ k      ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│     ┆ 17     ┆        ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│ 4   ┆ 2024-0 ┆ cualle ┆ WARNIN ┆ cds_lo ┆ is_com ┆ N/A   ┆ 4082 ┆ 0      ┆ 1.0   ┆ 1.0   ┆ PASS  │\n",
       "│     ┆ 6-16   ┆ e.chec ┆ G      ┆ cus_ta ┆ plete  ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│     ┆ 16:12: ┆ k      ┆        ┆ g      ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│     ┆ 17     ┆        ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│ 5   ┆ 2024-0 ┆ cualle ┆ WARNIN ┆ protei ┆ is_com ┆ N/A   ┆ 4082 ┆ 0      ┆ 1.0   ┆ 1.0   ┆ PASS  │\n",
       "│     ┆ 6-16   ┆ e.chec ┆ G      ┆ n_id   ┆ plete  ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│     ┆ 16:12: ┆ k      ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│     ┆ 17     ┆        ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│ 6   ┆ 2024-0 ┆ cualle ┆ WARNIN ┆ gene   ┆ is_uni ┆ N/A   ┆ 4082 ┆ 496    ┆ 0.878 ┆ 1.0   ┆ FAIL  │\n",
       "│     ┆ 6-16   ┆ e.chec ┆ G      ┆        ┆ que    ┆       ┆      ┆        ┆ 491   ┆       ┆       │\n",
       "│     ┆ 16:12: ┆ k      ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│     ┆ 17     ┆        ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│ 7   ┆ 2024-0 ┆ cualle ┆ WARNIN ┆ locus_ ┆ is_uni ┆ N/A   ┆ 4082 ┆ 0      ┆ 1.0   ┆ 1.0   ┆ PASS  │\n",
       "│     ┆ 6-16   ┆ e.chec ┆ G      ┆ tag    ┆ que    ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│     ┆ 16:12: ┆ k      ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│     ┆ 17     ┆        ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│ 8   ┆ 2024-0 ┆ cualle ┆ WARNIN ┆ cds_ge ┆ is_uni ┆ N/A   ┆ 4082 ┆ 496    ┆ 0.878 ┆ 1.0   ┆ FAIL  │\n",
       "│     ┆ 6-16   ┆ e.chec ┆ G      ┆ ne     ┆ que    ┆       ┆      ┆        ┆ 491   ┆       ┆       │\n",
       "│     ┆ 16:12: ┆ k      ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│     ┆ 17     ┆        ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│ 9   ┆ 2024-0 ┆ cualle ┆ WARNIN ┆ cds_lo ┆ is_uni ┆ N/A   ┆ 4082 ┆ 0      ┆ 1.0   ┆ 1.0   ┆ PASS  │\n",
       "│     ┆ 6-16   ┆ e.chec ┆ G      ┆ cus_ta ┆ que    ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│     ┆ 16:12: ┆ k      ┆        ┆ g      ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│     ┆ 17     ┆        ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│ 10  ┆ 2024-0 ┆ cualle ┆ WARNIN ┆ protei ┆ is_uni ┆ N/A   ┆ 4082 ┆ 45     ┆ 0.988 ┆ 1.0   ┆ FAIL  │\n",
       "│     ┆ 6-16   ┆ e.chec ┆ G      ┆ n_id   ┆ que    ┆       ┆      ┆        ┆ 976   ┆       ┆       │\n",
       "│     ┆ 16:12: ┆ k      ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "│     ┆ 17     ┆        ┆        ┆        ┆        ┆       ┆      ┆        ┆       ┆       ┆       │\n",
       "└─────┴────────┴────────┴────────┴────────┴────────┴───────┴──────┴────────┴───────┴───────┴───────┘"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from toolz import compose, first, juxt\n",
    "from toolz.curried import map as mapc\n",
    "from operator import attrgetter as at\n",
    "from operator import itemgetter as it\n",
    "from operator import methodcaller as mc\n",
    "from operator import eq\n",
    "from functools import partial\n",
    "import polars as pl\n",
    "from cuallee import Check\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import translate\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(action=\"ignore\")\n",
    "\n",
    "\n",
    "#filename = \"test/fixtures/unextected_features/FR823450_1.gb\"\n",
    "filename = \"temp/T01.gb\"\n",
    "genome = SeqIO.read(filename, \"genbank\")\n",
    "\n",
    "\n",
    "cds_qualifiers = (\"gene\", \"locus_tag\", \"protein_id\", \"function\", \"product\", \"translation\", \"transl_table\", \"codon_start\")\n",
    "gene_qualifiers = (\"gene\", \"locus_tag\")\n",
    "location_attributes = (\"start\", \"end\", \"strand\")\n",
    "\n",
    "annotations_attributes = (\"topology\", \"organism\", \"taxonomy\")\n",
    "topology, organism, taxonomy = compose(juxt(map(it, annotations_attributes)), at(\"annotations\"))(genome)\n",
    "\n",
    "id_attributes = (\"id\", \"name\", \"description\")\n",
    "id, name, description = juxt(map(at, id_attributes))(genome)\n",
    "\n",
    "_impute_attributes = lambda x: mc(\"get\", x, [\"\"])\n",
    "\n",
    "_type_cds = compose(partial(eq, \"CDS\"), at(\"type\"))\n",
    "_type_gene = compose(partial(eq, \"gene\"), at(\"type\"))\n",
    "\n",
    "data_cds = list(map(compose(list, mapc(first), juxt(map(_impute_attributes, cds_qualifiers)), at(\"qualifiers\")), filter(_type_cds, genome.features)))\n",
    "data_cds_pk = list(map(compose(list, mapc(at(\"real\")), juxt(map(at, location_attributes)), at(\"location\")), filter(_type_cds, genome.features)))\n",
    "\n",
    "df_cds = pl.DataFrame(data_cds, schema=cds_qualifiers)\n",
    "df_cds_pk = pl.DataFrame(data_cds_pk, schema=location_attributes)\n",
    "cds_extract = list(map(lambda x: str(x.extract(genome.seq)), filter(_type_cds, genome.features)))\n",
    "df_cds_extract = pl.DataFrame(cds_extract, schema=[\"extract\"])\n",
    "df_cds_translate = pl.DataFrame(map(partial(translate, stop_symbol=\"\", table=11), cds_extract), schema=[\"translation_fn\"])\n",
    "cds = pl.concat(items=[df_cds, df_cds_pk, df_cds_extract, df_cds_translate], how=\"horizontal\").rename(mapping={\"gene\": \"cds_gene\", \"locus_tag\": \"cds_locus_tag\"})\n",
    "\n",
    "\n",
    "data_gene = list(map(compose(list, mapc(first), juxt(map(_impute_attributes, gene_qualifiers)), at(\"qualifiers\")), filter(_type_gene, genome.features)))\n",
    "data_gene_pk = list(map(compose(list, mapc(at(\"real\")), juxt(map(at, location_attributes)), at(\"location\")), filter(_type_gene, genome.features)))\n",
    "\n",
    "df_gene = pl.DataFrame(data_gene, schema=gene_qualifiers)\n",
    "df_gene_pk = pl.DataFrame(data_gene_pk, schema=location_attributes)\n",
    "gene_extract = list(map(lambda x: str(x.extract(genome.seq)), filter(_type_gene, genome.features)))\n",
    "df_gene_translate = pl.DataFrame(map(partial(translate, stop_symbol=\"\", table=11), gene_extract), schema=[\"translation_fn\"])\n",
    "df_gene_extract = pl.DataFrame(gene_extract, schema=[\"extract\"])\n",
    "\n",
    "gene = pl.concat(items=[df_gene, df_gene_pk, df_gene_extract, df_gene_translate], how=\"horizontal\")\n",
    "\n",
    "\n",
    "df = cds.join(other=gene, on=[\"start\", \"end\", \"strand\", \"extract\"], how=\"full\", coalesce=True)\n",
    "df = df.with_columns(\n",
    "    id=pl.lit(id),\n",
    "    name=pl.lit(name),\n",
    "    description=pl.lit(description),\n",
    "    topology=pl.lit(topology),\n",
    "    organism=pl.lit(organism),\n",
    "    taxonomy=pl.lit(taxonomy),\n",
    "    filename=pl.lit(filename),\n",
    ").drop_nulls()\n",
    "\n",
    "\n",
    "# Create fasta file\n",
    "check = Check()\n",
    "check_columns = [\"gene\", \"locus_tag\", \"cds_gene\", \"cds_locus_tag\", \"protein_id\"]\n",
    "[check.is_complete(c) for c in check_columns]\n",
    "[check.is_unique(c) for c in check_columns]\n",
    "check.validate(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import uuid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
