# Navigating `synphage` pipeline


## Requirements 

???+ success "*Prerequisite:*"
    You need to have `synphage` installed in a python environment or in a docker container or to have pulled synphage docker image. Start synphage and open the Dagster UI in your browser to get started.  
    === "venv"
        ``` bash
        pip install synphage
        dagster dev -h 0.0.0.0 -p 3000 -m synphage
        ```
        For more details, see [installation instruction](installation.md#pip-install) or [how to run the software](installation.md/#run-synphage-pip).
    === "docker"
        ``` bash
        docker pull vestalisvirginis/synphage:<tag>
        docker run --rm --name my-synphage-container -p 3000 vestalisvirginis/synphage:<tag>
        ```
        For more details, see [installation instruction](installation.md/#docker-install) or [how to run the software](installation.md/#run-synphage-container).
    === "Docker Desktop"
        <iframe width="560" height="315"
        src="../images/phages/docker_desktop_movie.mp4" 
        frameborder="0" 
        allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" 
        allowfullscreen></iframe>  
        For more details, see [installation instruction](installation.md/#docker-install) or [how to run the software](installation.md/#run-synphage-container).


<figure markdown="span">
![Dagster landing page](./images/pipeline/dag_landing_page.png)
<figcaption>Dagster UI - landing page</figcaption>
</figure>


## Navigating the UI

To navigate to the jobs, go to 
![Image title](./images/phages/dagster-primary-mark.svg#only-light){ width="50"}
![Image title](./images/phages/dagster-reversed-mark.svg#only-dark){ width="50"} 
Dagster_home -> Jobs

<figure markdown="span">
![List of jobs](./images/pipeline/dag_jobs.png) 
<figcaption>Dagster shutting down</figcaption>
</figure>

Whenever you are lost, go back to the 
![Image title](./images/phages/dagster-primary-mark.svg#only-light){ width="50"}
![Image title](./images/phages/dagster-reversed-mark.svg#only-dark){ width="50"} to get back on track.


## Software structure

synphage pipeline is composed of `four steps` that need to be run `sequencially`:  
    - step 1: loading the data  
    - step 2: validating the data  
    - step 3: blasting the data  
    - step 4: ploting the data  


### Step 1: Loading the data into the pipeline

The step 1 is composed of two sub-steps : 
- `step_1a_get_user_data`, for loading user's data into the pipeline
- `step_1b_download`, for downloading data from the NCBI

GenBank files are loaded into the pipeline from the `input_folder` setup by the user `and/or` `downloaded` from the NCBI database.


#### step_1a: `step_1a_get_user_data`
    This job allows the users to upload their own data / data stored locally into the pipeline. (see [](installation.md#data_input_pip), [](installation.md#data_input_docker)).




### Job 4 : NCBI download job  

This job requires to have the `EMAIL` and `API_KEY` environnmental variable set in order to access the NCBI database.  

In order to set your keywords for the NCBI database query, click the arrow on the right side of the `materialise all` botton to access the drop-down menu and select `Open launchpad`.  

![Select launchpad for NCBI download job](./images/dagster/dag_job_4_launchpad.png)  
  
You can then set the `search_key` with your own keywords. Example: "Paenibacillus larvae"[Organism] AND complete genome[Title]  

![NCBI query configuration panel](./images/dagster/dag_job_4_launchpad_menu.png)  




### Step 2: Data validation
### Step 3: Blasting the data
### Step 4: Synteny plot


### Jobs

The current software is structured in four different jobs.  
 - `blasting_job` : create the blastn of each sequences against each sequences (results -> gene_identity folder)  
 - `transform` : create three tables from the blastn results and genbank files (results -> tables)  
 - `synteny_job` : create the synteny graph (results -> synteny)  
 - `ncbi_download_job` : download genomes to be analysed from the NCBI database (same combinaison of keywords can be used as in the ncbi website)  

???+ note
    Different synteny plots can be generated from the same set of genomes. In this case the two first jobs only need to be run once and the third job (`synteny_job`) can be triggered separately for each graphs.


### Output data

#### Tables 

The tables generated by the software are saved as parquet files. Those files can easily be read with a software such as Tad (https://duckdb.org/docs/archive/0.5.1/guides/data_viewers/tad).


#### Synteny plot  

The synteny plot is generated as `.svg` file and `.png` file, and contains the sequences indicated in the sequences .csv file. The genes are colour-coded according to their abundance (percentage) among the plotted sequences. The cross-links between each consecutive sequence indicates the percentage or similarities between those two sequences.
The `.svg` can be open with a software such as Inkscape (https://inkscape.org) and be further annotated if needed.  


##### Plotting config options  

Field Name | Description | Default Value
 ------- | ----------- | ----
`title` | Generated plot file title | synteny_plot
`colours` | Gene identity colour bar | ["#fde725", "#90d743", "#35b779", "#21918c", "#31688e", "#443983", "#440154"] 
`gradient` | Nucleotide identity colour bar | #B22222
`graph_shape` | Linear or circular representation | linear
`graph_pagesize` | Output document format | A4
`graph_fragments` | Number of fragments | 1
`graph_start` | Sequence start | 1
`graph_end` | Sequence end | length of the longest genome  


#### Genbank file download

The `ncbi_download_job` allow to download sequences of interest into the genbank folder to be subsequently processed by the software. 

##### Requirement  

Connection to the NCBI databases requires user's `email` and `api_key`.
```bash
export EMAIL=user.email@email.com
export API_KEY=UserApiKey
```  

##### Query config options  

Field Name | Description | Default Value
 ------- | ----------- | ----
`search_key` | Keyword(s) for NCBI query | Myoalterovirus
`database` | Database identifier | nuccore  



### Job 1 : blasting job

The first step of the pipeline is to run blastn between each genomes using the blast+ tool from the ncbi (https://doi.org/10.1186/1471-2105-10-421). This job require that sequences have been added to the genbank folder. In order to run the job, hit the rigth up corner botton labelled `Materialize all` .

![Job 1 pipeline](./images/dagster/dag_job_1.png) 

![Job 1 running](./images/dagster/dag_job_1_launched.png) 

Once an element is indicated as materialized, metadata related to the job will be available in the rigth panel. See example below, showing the list of the sequence that have been processed. 

![Job 1 finished : all the elements were materialized](./images/dagster/dag_job_1_finished.png) 

Once all the assets are materialized, new sequences can be added to the folder and be processed the same way.

### Job 2 : transform job <a id="step2-validation-job"></a>

![Job 2 pipeline](./images/dagster/dag_job_2.png) 

In order to run the second job, genbank files needs to be present in the genbank folder and the first job needs to have been materialized.  
To run run the job, hit the Launchpad tab and and then the rigth down-corner botton `Launch run`.

![Job 2 Launchpad](./images/dagster/dag_job_2_launcher.png) 

As previously you can follow the progression of the job.

![Progression of the job 2 run](./images/dagster/dag_job_2_launched.png) 

![Job 2 finished](./images/dagster/dag_job_2_finished.png) 

Job 2 needs to be re-run after Job 1 in order to integrate the new data to the table.


### Job 3 : synteny job


    ???+ warning
        Please here use **only** `.gb` as file extension.

    ???+ info
        The integer after the comma represents the orientation of the sequence in the synteny diagram.
        0 : sequence
        1 : reverse


        8. For ploting add a `sequences.csv` file in the /data directory. Please use the file editor of the docker to check that the format of your file is according to the example below:
    ```txt
    168_SPbeta.gb,0
    Phi3T.gb,1
    ```
    Example of incorrectly formatted csv file (can happen when saved from excel):
    ![Incorrectly formatted csv file](./images/dd_csv_file_excel.png){align=right}
    Example of correctly formatted csv file:  
    ![Correctly formatted csv file](./images/dd_csv_file_correctly_formatted.png){align=right}

        ???+ warning
            Please here use **only** `.gb` as file extension.

        ???+ info
            The integer after the comma represents the orientation of the sequence in the synteny diagram.
            0 : sequence
            1 : reverse


             5. For ploting add a sequences.csv file in the /data directory. Format your file according to the example below:
    ```txt
    168_SPbeta.gb,0
    Phi3T.gb,1
    ```
    ```bash
    # Create file
    touch sequences.csv
    # Edit file
    vim sequences.csv
    # Copy file to the /data directory
    docker cp path_to_file/sequences.csv /data/
    ```

        ???+ warning
            Please here use **only** `.gb` as file extension.

        ???+ info
            The integer after the comma represents the orientation of the sequence in the synteny diagram.
            0 : sequence
            1 : reverse

This job requires the genbank files to be available in the genbank folder, that job 1 and 2 have been run on the sequence of interest and that a sequences.csv is present in the /data folder.

As for the first job, this job is triggered by hiting the `Materialize all` botton.

![Job 3 pipeline](./images/dagster/dag_job_3.png) 

In the same way as previously, you can follow the progression of the job.

![Job 3 run progression](./images/dagster/dag_job_3_finished.png) 

Metadata for each assets are also available in the rigth panel for each asset. 

![Job 3 plot metadata](./images/dagster/dag_job_3_metadata.png) 

The `sequences.csv` file can be amended and the job can be run again (2 more sequences were added for the plot):

![Job 3 with 4 sequences in the csv file](./images/dagster/dag_job_3_finished_2.png) 

![New run metadata](./images/dagster/dag_job_3_metadata_2.png) 

In order to change the configurations for the graphic, click the arrow on the right side of the `materialise all` botton to access the drop-down menu and select `Open launchpad`.

![Select launchpad](./images/dagster/dag_job_3_launchpad.png) 

In the launchpad it is possible to replace any config values by your own configuration. For example, the value for the gradient can be changed from '#B22222' (red brick colour) to '#c0c0c0' (silver grey colour). The list of config options is defined in the 'Synteny plot --> List of config options' sections.

![Graphic configuration panel](./images/dagster/dag_job_3_launchpad_menu.png) 




### Metadata and results of each run

Dagster wil keep track of the job that have been run and of he outcome.

The informations can be found in the `Assets` panel.

![Assets](./images/dagster/dag_assets_summary.png) 

For each asset, you can review the metadata generated during the run as for the below example regarding the creation of the synteny plot.

![Example asset: create_graph](./images/dagster/dag_graph_metadata.png)  




local_resource:
    config:
      input_dir: temp/sequences_2
      output_dir: /data